sympy/polys/matrices/lll.py

```diff
@@ -0,0 +1,135 @@
+    .. [2] https://web.cs.elte.hu/~lovasz/scans/lll.pdf
+    .. [3] Murray R. Bremner, "Lattice Basis Reduction: An Introduction to the LLL Algorithm and Its Applications"
+
+    """
+    assert QQ(1, 4) < delta < QQ(1, 1), "delta must lie in range (0.25, 1)"
+    assert x.shape[0] <= x.shape[1], "input matrix must have shape (m, n) with m <= n"
+    assert isinstance(x.domain, IntegerRing), "matrix domain must be IntegerRing subclass"
+    m = x.shape[0]
+    n = x.shape[1]
+    k = 1
+    y = x.copy()
+    y_star = x.zeros((m, n), QQ)
+    mu = x.zeros((m, m), QQ)
+    g_star = [QQ(0, 1) for _ in range(m)]
+    half = QQ(1, 2)
+
+    def closest_integer(x):
+        return ZZ(mfloor(x + half))
+
+    def lovasz_condition(k: int) -> bool:
+        return g_star[k] >= ((delta - pow(mu[k][k - 1], 2)) * g_star[k - 1])
+
+    def mu_small(k: int, j: int) -> bool:
+        return abs(mu[k][j]) <= half
+
+    def dot_rows(x, y, rows: Tuple[int, int]):
+        return sum([x[rows[0]][z] * y[rows[1]][z] for z in range(x.shape[1])])
+
+    def reduce_row(mu, y, rows: Tuple[int, int]):
+        r = closest_integer(mu[rows[0]][rows[1]])
+        y[rows[0]] = [y[rows[0]][z] - r * y[rows[1]][z] for z in range(n)]
+        for j in range(rows[1]):
+            mu[rows[0]][j] -= r * mu[rows[1]][j]
+        mu[rows[0]][rows[1]] -= r
+
+    try:
+        for i in range(m):
+            y_star[i] = [QQ.convert_from(z, ZZ) for z in y[i]]
+            for j in range(i):
+                mu[i][j] = dot_rows(y, y_star, (i, j)) / g_star[j]
+                y_star[i] = [y_star[i][z] - mu[i][j] * y_star[j][z] for z in range(n)]
+            g_star[i] = dot_rows(y_star, y_star, (i, i))
+        while k < m:
+            if not mu_small(k, k - 1):
+                reduce_row(mu, y, (k, k - 1))
+            if lovasz_condition(k):
+                for l in range(k - 2, -1, -1):
+                    if not mu_small(k, l):
+                        reduce_row(mu, y, (k, l))
+                k += 1
+            else:
+                nu = mu[k][k - 1]
+                alpha = g_star[k] + pow(nu, 2) * g_star[k - 1]
+                beta = g_star[k - 1] / alpha
+                mu[k][k - 1] = nu * beta
+                g_star[k] = g_star[k] * beta
+                g_star[k - 1] = alpha
+                y[k], y[k - 1] = y[k - 1], y[k]
+                for j in range(k - 1):
+                    mu[k][j], mu[k - 1][j] = mu[k - 1][j], mu[k][j]
+                for i in range(k + 1, m):
+                    xi = mu[i][k]
+                    mu[i][k] = mu[i][k - 1] - nu * xi
+                    mu[i][k - 1] = mu[k][k - 1] * mu[i][k] + xi
+                k = max(k - 1, 1)
+        assert all([lovasz_condition(i) for i in range(1, m)])
+        assert all([mu_small(i, j) for i in range(m) for j in range(i)])
+        return y
+    except ZeroDivisionError:
+        raise DMRankError("input matrix contains linearly dependent rows")
```

Comment:
In general try/except should be used around the smallest block of code possible to avoid catching an exception from the wrong place. There are only two lines in the function above that use division so the try except should be just around one of those or if it needs to be in both places then there should be two separate try/except blocks. Actually it would be better just to test for zero like:
```
if not x:
    raise DMRankError
```
Would it be difficult to handle this by permuting the zero rows rather than raising an error?

Reply:
I think if the rows are linearly dependent it will produce a zero division error no matter what we do, the algorithm is designed to zero some rows out in that case. So I will try/catch both of the divisions to make sure I catch this. I don't know if it is only one which can produce it, it looks like it could be both spots which can do it.